cmd/fs-v1-metadata.go

```diff
@@ -274,167 +260,83 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
-	lk, err := lock.LockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
-	if err != nil {
-		return traceError(err)
-	}
-	defer lk.Close()
-
-	var format = &formatConfigV1{}
-	_, err = format.ReadFrom(lk)
-	// For all unexpected errors, we return.
-	if err != nil && errorCause(err) != io.EOF {
-		return traceError(fmt.Errorf("Unable to load 'format.json', %s", err))
-	}
-
-	// If we couldn't read anything, The disk is unformatted.
-	if errorCause(err) == io.EOF {
-		err = errUnformattedDisk
-		format = newFSFormatV2()
-	} else {
-		// Validate loaded `format.json`.
-		err = checkFormatFS(format, fsFormatVersion)
-		if err != nil && err != errFSFormatOld {
-			return traceError(fmt.Errorf("Unable to validate 'format.json', %s", err))
-		}
-	}
-
-	// Disk is in old format migrate object metadata.
-	if err == errFSFormatOld {
-		if merr := migrateFSObject(fsPath, fsUUID); merr != nil {
-			return merr
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, formatConfigFile)
+
+	// This loop validates format.json by holding a read lock and
+	// proceeds to holding a TryWriteLock by validating if the
+	// disk is unformatted. If for some reason TryWriteLock fails
+	// and the error returned is lock.ErrAlreadyLocked i.e some other
+	// process is holding a lock we make sure that we try again
+	// in the loop by holding a read lock.
+	for {
+		var rlk *lock.RLockedFile
+		rlk, err = checkLockedValidFormatFS(fsPath)
+		if err == nil {
+			// Format is valid, close the read lock.
+			rlk.Close()
 		}
-
-		// Initialize format v2.
-		format = newFSFormatV2()
-	}
-
-	// Rewrite or write format.json depending on if disk
-	// unformatted and if format is old.
-	if err == errUnformattedDisk || err == errFSFormatOld {
-		if _, err = format.WriteTo(lk); err != nil {
-			return traceError(fmt.Errorf("Unable to initialize 'format.json', %s", err))
+		if err != nil && errorCause(err) == errUnformattedDisk {
+			// Attempt a write lock on formatConfigFile `format.json`
+			// file stored in minioMetaBucket(.minio.sys) directory.
+			lk, lerr := lock.TryLockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
+			if lerr != nil {
+				// Existing write locks detected.
+				if lerr == lock.ErrAlreadyLocked {
+					// Retry after waiting 100ms to avoid 100% CPU usage.
+					time.Sleep(100 * time.Millisecond)
+					// Continue and attempt write lock again.
+					continue
+				}
+				err = traceError(lerr)
+				break
+			}
+			defer lk.Close()
+
+			// Load format to validate again.
+			var format = &formatConfigV1{}
+			if lerr = format.LoadFormat(lk); lerr != nil && errorCause(lerr) == errUnformattedDisk {
+				if _, lerr = newFSFormat().WriteTo(lk); lerr != nil {
+					err = traceError(fmt.Errorf("Unable to initialize 'format.json', %s", lerr))
+					break
+				}
+			}
+			err = lerr
 		}
+		break
 	}
-
-	// Check for sanity.
-	return checkFormatSanityFS(fsPath, format.FS.Version)
+	return
 }
```

Comment:
This approach for `initFormatFS()` requires at-least two-locks in the best case. In the best case, the disk is already formatted (and is of the latest version, when V2 is introduced). This implementation takes a read lock inside `initFormatFS()` and then superfluously releases it, and then after `initFormatFS()` acquires and keeps the lock for the lifetime of the process.

We can do better than this.

Here is some code illustrating the idea of a state machine implementation that acquires locks and performs actions (formatting/migration) at well defined conditions and finally when the loop is exited, the disk is formatted (and in future V2 servers, migrations are also performed). The read lock is acquired just once in the best case, and held till the end of the process life-time.

```diff
diff --git a/cmd/fs-v1-metadata.go b/cmd/fs-v1-metadata.go
index 4be01ec..90a4eb0 100644
--- a/cmd/fs-v1-metadata.go
+++ b/cmd/fs-v1-metadata.go
@@ -25,6 +25,7 @@ import (
 	pathutil "path"
 	"sort"
 	"strings"
+	"time"
 
 	"github.com/minio/minio/pkg/lock"
 	"github.com/minio/minio/pkg/mimedb"
@@ -290,50 +291,67 @@ func checkLockedValidFormatFS(fsPath string) (*lock.RLockedFile, error) {
 	return rlk, traceError(err)
 }
 
+var errOldFormat = fmt.Errorf("Version is not latest!")
+
 func initFormatFS(fsPath, fsUUID string) (err error) {
 	fsFormatPath := pathJoin(fsPath, minioMetaBucket, formatConfigFile)
 
-	// This loop validates format.json by holding a read lock and
-	// proceeds to holding a TryWriteLock by validating if the
-	// disk is unformatted. If for some reason TryWriteLock fails
-	// and the error returned is lock.ErrAlreadyLocked i.e some other
-	// process is holding a lock we make sure that we try again
-	// in the loop by holding a read lock.
+	var rlk *lock.RLockedFile
 	for {
-		var rlk *lock.RLockedFile
 		rlk, err = checkLockedValidFormatFS(fsPath)
-		if err == nil {
-			// Format is valid, close the read lock.
-			rlk.Close()
-		}
-		if err != nil && errorCause(err) == errUnformattedDisk {
-			// Attempt a write lock on formatConfigFile `format.json`
-			// file stored in minioMetaBucket(.minio.sys) directory.
+		switch {
+		case err == nil:
+			// Disk is formatted, and has latest
+			// version. Hold the lock for lifetime of the
+			// process and exit initialization.
+			_ = rlk
+			return err
+		case errCause(err) == errUnformattedDisk || errCause(err) == errOldFormat:
+			// format.json does not exist or is of old version
 			lk, lerr := lock.TryLockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
-			if lerr != nil {
-				// Existing write locks detected.
-				if lerr == lock.ErrAlreadyLocked {
-					// Continue and attempt write lock again.
-					continue
-				}
-				err = traceError(lerr)
-				break
+			if lerr == lock.ErrAlreadyLocked {
+				// lock is already taken - sleep and retry
+				time.Sleep(100 * time.Millisecond)
+				continue
+			} else if lerr != nil {
+				// unexpected error - quit
+				return traceError(lerr)
 			}
+
+			// lock is acquired!
 			defer lk.Close()
 
-			// Load format to validate again.
+			// Load format - and check if someone
+			// formatted/migrated before we got the lock!
 			var format = &formatConfigV1{}
-			if lerr = format.LoadFormat(lk); lerr != nil && errorCause(lerr) == errUnformattedDisk {
+			if lerr = format.LoadFormat(lk); errorCause(lerr) == errUnformattedDisk {
 				if _, lerr = newFSFormat().WriteTo(lk); lerr != nil {
-					err = traceError(fmt.Errorf("Unable to initialize 'format.json', %s", lerr))
-					break
+					return traceError(fmt.Errorf("Unable to initialize 'format.json', %s", lerr))
 				}
+				// formatting worked fine. Acquire
+				// read-lock and exit init
+				continue
+			} else if lerr != nil {
+				// unexpected error - exit.
+				return traceError(lerr)
+			}
+
+			// validate the loaded format.
+			if lerr = format.CheckFS(); lerr == errOldFormat {
+				// we have write lock - perform migration!
+
+				// currently just a no-op. after
+				// migration, acquire read lock and
+				// exit init.
+			} else if lerr != nil {
+				// unexpected error - exit.
+				return traceError(lerr)
 			}
-			err = lerr
+		default:
+			// unexpected error received.
+			return traceError(err)
 		}
-		break
 	}
-	return
 }
 
 // Return if the part info in uploadedParts and completeParts are same.
diff --git a/cmd/fs-v1.go b/cmd/fs-v1.go
index 94e1dbf..6f9d0a9 100644
--- a/cmd/fs-v1.go
+++ b/cmd/fs-v1.go
@@ -127,24 +127,6 @@ func newFSObjectLayer(fsPath string) (ObjectLayer, error) {
 		return nil, err
 	}
 
-	// Validate the `format.json` again after the initialization
-	// to further validate `format.json` entries to be expected.
-	//
-	// Once the filesystem has initialized hold the read lock for
-	// the life time of the server. This is done to ensure that under
-	// shared backend mode for FS, remote servers do not migrate
-	// or cause changes on backend format after initFormatFS().
-	// There is a slight window for this to happen.
-	rlk, err := checkLockedValidFormatFS(fsPath)
-	if err != nil {
-		return nil, err
-	}
-
-	// Holding a read lock ensures that any write lock operation
-	// is blocked if attempted in-turn avoiding corruption on
-	// the backend disk.
-	_ = rlk // Hold the lock on `format.json` until server dies.
-
 	// Initialize and load bucket policies.
 	err = initBucketPolicies(fs)
 	if err != nil {
```